source/opt/copy_prop_arrays.cpp

```diff
@@ -0,0 +1,562 @@
+
+  if (!HasNoStores(source->GetVariable())) {
+    return nullptr;
+  }
+  return source;
+}  // namespace opt
+
+void CopyPropagateArrays::PropagateObject(ir::Instruction* var_inst,
+                                          MemoryObject* source) {
+  assert(var_inst->opcode() == SpvOpVariable &&
+         "This funciton propagtages variables.");
+
+  ir::Instruction* insertion_point = var_inst->NextNode();
+  while (insertion_point->opcode() == SpvOpVariable) {
+    insertion_point = insertion_point->NextNode();
+  }
+  ir::Instruction* new_access_chain =
+      BuildNewAccessChain(insertion_point, source);
+  context()->KillNamesAndDecorates(var_inst);
+  UpdateUses(var_inst, new_access_chain);
+}
+
+ir::Instruction* CopyPropagateArrays::BuildNewAccessChain(
+    ir::Instruction* insertion_point,
+    CopyPropagateArrays::MemoryObject* source) const {
+  analysis::ConstantManager* const_mgr = context()->get_constant_mgr();
+
+  InstructionBuilder builder(context(), insertion_point,
+                             ir::IRContext::kAnalysisDefUse |
+                                 ir::IRContext::kAnalysisInstrToBlockMapping);
+
+  analysis::Integer int_type(32, false);
+  const analysis::Type* uint32_type =
+      context()->get_type_mgr()->GetRegisteredType(&int_type);
+
+  // Convert the access chain in the source to a series of ids that can be used
+  // by the |OpAccessChain| instruction.
+  std::vector<uint32_t> index_ids;
+  for (uint32_t index : source->AccessChain()) {
+    const analysis::Constant* index_const =
+        const_mgr->GetConstant(uint32_type, {index});
+    index_ids.push_back(
+        const_mgr->GetDefiningInstruction(index_const)->result_id());
+  }
+
+  // Get the type for the result of the OpAccessChain
+  uint32_t pointer_type_id = source->GetPointerTypeId();
+
+  // Build the access chain instruction.
+  return builder.AddAccessChain(pointer_type_id,
+                                source->GetVariable()->result_id(), index_ids);
+}
+
+bool CopyPropagateArrays::HasNoStores(ir::Instruction* ptr_inst) {
+  return get_def_use_mgr()->WhileEachUser(
+      ptr_inst, [this](ir::Instruction* use) {
+        if (use->opcode() == SpvOpLoad) {
+          return true;
+        } else if (use->opcode() == SpvOpAccessChain) {
+          return HasNoStores(use);
+        } else if (use->IsDecoration() || use->opcode() == SpvOpName) {
+          return true;
+        } else if (use->opcode() == SpvOpStore) {
+          return false;
+        }
+        // Some other instruction.  Be conservative.
+        return false;
+      });
+}
+
+bool CopyPropagateArrays::HasValidReferencesOnly(ir::Instruction* ptr_inst,
+                                                 ir::Instruction* store_inst) {
+  ir::BasicBlock* store_block = context()->get_instr_block(store_inst);
+  opt::DominatorAnalysis* dominator_analysis =
+      context()->GetDominatorAnalysis(store_block->GetParent(), *cfg());
```

Comment:
Optional.  Move this to a class instance variable and call it once per function?  Saves the constant lookup in GetDominatorAnalysis.  OTOH, we may not even need to compute dominance if there are no copy-prop opportunities, so... *shrug*.

Reply:
I'll leave this as it for now.  Because it is just a lookup.  We could do a local cache or something, but not worth the complication for now.